File: java/org/simple/lifeiseasy/HigherOrderFunctions.java
package org.simple.lifeiseasy;

public class HigherOrderFunctions {

	public static void main(String[] args) {
		HigherOrderFunctions obj = new HigherOrderFunctions();
		System.out.println("sum of Numbers between 1 and 5 = " + obj.sumOfNumbers(1, 5));
		System.out.println("sum of Cubes between 1 and 5 = " + obj.sumOfCubes(1, 5));
		System.out.println("sum of Factorials between 1 and 5 = " + obj.sumOfFactorials(1, 5));
	}

	public int sumOfNumbers(int a, int b) {
		if (a > b) {
			return 0;
		} else {
			return a + sumOfNumbers(a + 1, b);
		}
	}

	public int sumOfCubes(int a, int b) {
		if (a > b) {
			return 0;
		} else {
			return a * a * a + sumOfCubes(a + 1, b);
		}
	}

	public int sumOfFactorials(int a, int b) {
		if (a > b) {
			return 0;
		} else {
			return fact(a) + sumOfFactorials(a + 1, b);
		}
	}

	public int fact(int a) {
		if (a <= 1) {
			return a;
		} else {
			return a * fact(a - 1);
		}
	}

}

File: scala/org/simple/lifeiseasy/HigherOrderFunctions.scala
package org.simple.lifeiseasy

import sun.applet.Main

object HigherOrderFunctionsObj {
  
  def main(args: Array[String]): Unit = {
    val sumOfNumbersBetween1and5 = sumOfNumbers(1,5);
    println(s"sum of numbers between 1 and 5 =$sumOfNumbersBetween1and5")
    val sumOfCubesBetween1and5 = sumOfCubes(1,5);
    println(s"sum of cubes between 1 and 5 =$sumOfCubesBetween1and5")
    val sumOfFactorialsBetween1and5 = sumOfFactorials(1,5);
    println(s"sum of factorials between 1 and 5 =$sumOfFactorialsBetween1and5")
  }
  
  def sumOfNumbers(a: Int, b: Int): Int =
    if(a>b) 0 else a + sumOfNumbers(a+1, b)
    
  def sumOfCubes(a: Int, b: Int): Int =
    if(a>b) 0 else a*a*a + sumOfCubes(a+1, b)
    
  def sumOfFactorials  (a: Int, b: Int): Int =
    if(a>b) 0 else fact(a) + sumOfFactorials(a+1, b)
    
  def fact(a: Int): Int = 
    if(a<=1) a else a * fact(a-1)
}